src/ripple/app/tx/impl/TxQImpl.cpp

```diff
@@ -0,0 +1,969 @@
+    getTxnsExpected() const
+    {
+        std::lock_guard <std::mutex> sl(lock_);
+
+        return txnsExpected_;
+    }
+
+    std::uint32_t
+    getEscalationMultiplier() const
+    {
+        std::lock_guard <std::mutex> sl(lock_);
+
+        return escalationMultiplier_;
+    }
+
+    std::uint64_t
+    scaleFeeLevel(OpenView const& view) const;
+};
+
+CandidateTxn::CandidateTxn(
+    std::shared_ptr<STTx const> const& txn,
+        TxID const& txID, std::uint64_t const& feeLevel,
+            ApplyFlags const flags,
+                PreflightResult const& pfresult)
+    : txn_(txn)
+    , feeLevel_(feeLevel)
+    , txID_(txID)
+    , account_(txn->getAccountID(sfAccount))
+    , sequence_(txn->getSequence())
+    , flags_(flags)
+    , pfresult_(pfresult)
+{
+    if (txn->isFieldPresent(sfLastLedgerSequence))
+        lastValid_ = txn->getFieldU32(sfLastLedgerSequence);
+
+    if (txn->isFieldPresent(sfAccountTxnID))
+        priorTxID_ = txn->getFieldH256(sfAccountTxnID);
+}
+
+TxQAccount::TxQAccount(std::shared_ptr<STTx const> const& txn)
+    :TxQAccount(txn->getAccountID(sfAccount))
+{
+}
+
+TxQAccount::TxQAccount(const AccountID& account)
+    : account_(account)
+    , totalFees_(0)
+{
+}
+
+CandidateTxn&
+TxQAccount::addCandidate(CandidateTxn&& txn)
+{
+    auto fee = txn.getFeeLevel();
+    totalFees_ += fee;
+    auto sequence = txn.getSequence();
+
+    auto result = transactions_.emplace(std::make_pair(sequence, std::move(txn)));
+    assert(result.second);
+    assert(&result.first->second != &txn);
+
+    return result.first->second;
+}
+
+bool
+TxQAccount::removeCandidate(TxSeq const& sequence)
+{
+    return transactions_.erase(sequence) != 0;
+}
+
+CandidateTxn const*
+TxQAccount::findCandidateAt(TxSeq const& sequence) const
+{
+    auto iter = transactions_.find(sequence);
+    return iter == transactions_.end() ?
+        nullptr : &iter->second;
+}
+
+std::size_t
+FeeMetrics::updateFeeMetrics(Application& app,
+    ReadView const& view, bool timeLeap)
+{
+    std::vector<uint64_t> feeLevels;
+    for (auto const& tx : view.txs)
+    {
+        auto const baseFee = calculateBaseFee(app, view,
+            *tx.first, j_);
+        feeLevels.push_back(getFeeLevelPaid(*tx.first,
+            baseLevel, baseFee));
+    }
+    std::sort(feeLevels.begin(), feeLevels.end());
+    auto const size = feeLevels.size();
+
+    std::lock_guard <std::mutex> sl(lock_);
+
+    JLOG(j_.debug) << "Ledger " << view.info().seq <<
+        " has " << size << " transactions. " <<
+        "Ledgers are processing " <<
+        (timeLeap ? "slowly" : "as expected") <<
+        ". Expected transactions is currently " <<
+        txnsExpected_ << " and multiplier is " <<
+        escalationMultiplier_;
+
+    if (timeLeap)
+    {
+        // Ledgers are taking to long to process,
+        // so clamp down on limits.
+        txnsExpected_ = std::min(std::max(
+            feeLevels.size(), minimumTxnCount_),
+            targetTxnCount_ - 1);
```

Comment:
Could use `boost::clamp(feeLevels.size(), minimumTxnCount_, targetTxnCount_-1)`. But really fine as is (and no one has to look up what `boost::clamp' is. Your call.